src/ConfigurationClientManager.ts

```diff
@@ -0,0 +1,332 @@
+            this.#lastFallbackClientRefreshAttempt = currentTime;
+            const url = new URL(this.endpoint);
+            await this.#discoverFallbackClients(url.hostname);
+        }
+    }
+
+    async #discoverFallbackClients(host) {
+        const timeout = setTimeout(() => {
+        }, SrvQueryTimeout);
+        const srvResults = await querySrvTargetHost(host);
+
+        try {
+            const result = await Promise.race([srvResults, timeout]);
+
+            if (result === timeout) {
+                throw new Error("SRV record query timed out.");
+            }
+
+            const srvTargetHosts = result as string[];
+            // Shuffle the list of SRV target hosts
+            for (let i = srvTargetHosts.length - 1; i > 0; i--) {
+                const j = Math.floor(Math.random() * (i + 1));
+                [srvTargetHosts[i], srvTargetHosts[j]] = [srvTargetHosts[j], srvTargetHosts[i]];
+            }
+
+            const newDynamicClients: ConfigurationClientWrapper[] = [];
+            for (const host of srvTargetHosts) {
+                if (isValidEndpoint(host, this.#validDomain)) {
+                    const targetEndpoint = `https://${host}`;
+                    if (targetEndpoint.toLowerCase() === this.endpoint.toLowerCase()) {
+                        continue;
+                    }
+                    const client = this.#newConfigurationClient(targetEndpoint);
+                    newDynamicClients.push(new ConfigurationClientWrapper(targetEndpoint, client));
+                }
+            }
+
+            this.#dynamicClients = newDynamicClients;
+            this.#lastFallbackClientRefreshTime = Date.now();
+        } catch (err) {
+            console.warn(`Fail to build fallback clients, ${err.message}`);
+        }
+    }
+
+    #newConfigurationClient(endpoint) {
+        if (this.#credential) {
+            return new AppConfigurationClient(endpoint, this.#credential, this.#clientOptions);
+        }
+
+        const connectionStr = buildConnectionString(endpoint, this.#secret, this.#id);
+        return new AppConfigurationClient(connectionStr, this.#clientOptions);
+    }
+
+    #isFallbackClientDiscoveryDue(dateTime) {
+        return dateTime >= this.#lastFallbackClientRefreshAttempt + MinimalClientRefreshInterval
+            && (!this.#dynamicClients
+                || this.#dynamicClients.every(client => dateTime < client.backoffEndTime)
+                || dateTime >= this.#lastFallbackClientRefreshTime + FallbackClientRefreshExpireInterval);
+    }
+}
+
+/**
+ * Query SRV records and return target hosts.
+ * @param {string} host - The host to query.
+ * @returns {Promise<string[]>} - A promise that resolves to an array of target hosts.
+ */
+async function querySrvTargetHost(host) {
+    const results: string[] = [];
+    let dns;
+
+    if (isFailoverableEnv()) {
+        dns = require("dns/promises");
+    } else {
+        return results;
```

Comment:
We actually don't expect to execute this method if it's not a failoverable environment right? Then let's throw exception here, rather than return an empty list.